sip-quickstart/assets/admin/actions.private.js

```diff
@@ -0,0 +1,205 @@
+      friendlyName,
+    });
+    const authCallsMapping = this.client.sip
+      .domains(sipDomainSid)
+      .auth.calls.credentialListMappings.create({
+        credentialListSid: credentialList.sid,
+      });
+    const domainRegistrationMapping = this.client.sip
+      .domains(sipDomainSid)
+      .auth.registrations.credentialListMappings.create({
+        credentialListSid: credentialList.sid,
+      });
+    await Promise.all([authCallsMapping, domainRegistrationMapping]);
+    await this.client.sip.domains(sipDomainSid).update({
+      sipRegistration: true,
+    });
+    return {
+      CREDENTIAL_LIST_SID: credentialList.sid,
+    };
+  }
+
+  async createDefaultCredentialListForDomain({ sipDomainSid }) {
+    const serviceName = this.options.DOMAIN_NAME.replace(".twil.io", "");
+    const friendlyName = `${this.options.friendlyName} (${serviceName})`;
+    const results = await this.createMappedCredentialList({
+      friendlyName,
+      sipDomainSid,
+    });
+    await this.addDefaultCredentials({
+      credentialListSid: results.CREDENTIAL_LIST_SID,
+    });
+    return results;
+  }
+
+  async addDefaultCredentials({ credentialListSid }) {
+    const usernames = extensions.map((ext) => ext.username);
+    await this.addNewCredentials({ credentialListSid, usernames });
+  }
+
+  async addNewCredential({ credentialListSid, username, password }) {
+    if (password === undefined) {
+      password = process.env.DEFAULT_SIP_USER_PASSWORD;
+    }
+    await this.client.sip
+      .credentialLists(credentialListSid)
+      .credentials.create({ username, password });
+  }
+
+  async addNewCredentials({ credentialListSid, usernames }) {
+    const promises = usernames.map((username) =>
+      this.addNewCredential({
+        credentialListSid,
+        username,
+      })
+    );
+    await Promise.all(promises);
+  }
+
+  async updateSipDomainVoiceUrl({ sipDomainSid, voiceUrl }) {
+    await this.client.sip.domains(sipDomainSid).update({ voiceUrl });
+  }
+
+  async chooseLogicalIncomingNumber() {
+    const incomingNumbers = await this.client.incomingPhoneNumbers.list();
+    return incomingNumbers.find(
+      (number) => number.voiceUrl === DEFAULT_TWILIO_WEBHOOK
+    );
+  }
+
+  async chooseLogicalCallerId() {
+    const incomingNumbers = await this.client.incomingPhoneNumbers.list({
+      limit: 1,
+    });
+    let number;
+    if (incomingNumbers.length > 0) {
+      number = incomingNumbers[0].phoneNumber;
+    } else {
+      const outgoingCallerIds = await this.client.outgoingCallerIds.list({
+        limit: 1,
+      });
+      // Required right?
+      number = outgoingCallerIds[0].phoneNumber;
+    }
+    return number;
+  }
+
+  async updateIncomingNumber({ sid, voiceUrl }) {
+    const incomingNumber = await this.client.incomingPhoneNumbers(sid).fetch();
+    await incomingNumber.update({ voiceUrl });
+    return {
+      INCOMING_NUMBER: incomingNumber.phoneNumber,
+    };
+  }
+
+  async clearIncomingNumber() {
```

Comment:
This and the two functions below don't need to be asynchronous, so shouldn't be marked as such and shouldn't be called with await when they are called.

But, I'm guessing this is done like this because they can be called in the same method as all the other async methods and this makes the meta programming easier. If that's the case, I would think you can ignore me!

Reply:
Yeah they are expected to be promises from the callers...agreed that it's kind of dorky. I might be able to make the runner more lenient. Will let you know!